addons/purchase/tests/test_purchase_invoice.py

```diff
@@ -814,7 +814,7 @@ def test_subset_match_from_edi_and_diff_unit_price(self):
         invoice_lines = invoice.line_ids.filtered(lambda l: l.price_unit)
         self.assertEqual(len(invoice_lines), 2)
         for line in invoice_lines:
-            if (line.product_id == self.product_order):
+            if (line.product_id.name != self.product_order.name):
```

Comment:
Okay so this one is kinda complicated so sorry in advance if the explanation is messy :sweat_smile: 

This change has to be done as a result from the invoice_line_ids.name being empty most of the time now.
In the purchase app, there is a mechanism that handles matching invoices with purchase orders and it takes some arbitrary decisions that make little sense to me but that another problem. For example, the matching is done only if the sum of the invoice's lines' amount to invoice is within a 0.02 units of the currency from the po total amount which is weird. This means that a change like emptying the line.name can end up making the matching not match the same invoices and purchase orders :upside_down_face: And i'm not talking about code, i litteraly mean having two identical invoices with multiple lines expect for one with empty line.name(s) and two identical po. They should match the same way with one another but it might not be the case.

This test is a scenario where it happens. The po has two lines which have a product. Let's say price X and X and name A and B (both product have the same price). The invoice has four line, two of which have a product of prices X and X but the names are important. While before they would be named B and A, now the line names are empty. So as it would check for a matching name and find that the first line of the po has the same name as the second from the invoice, they would match. Now that the names are empty, when it tries to match the first line of the po, it doesn't find any line in the invoice with the same name so it tkaes the first line with same amount which is the first line. and then matched the second with the second doing the same process.

If you want to check by yourself, the process goes trough multiple methods that are all more complicated and arbitrary than the other imo but the real import line happens in `/home/odoo/dev/src/odoo/addons/purchase/models/account_invoice.py` at line 260

It doesn't change anything important because many checks are done before matching. The semantic of this test isn't changed either, the only change is that the product on the invoice_line matched with the po is the other product with the same price. In real scenarios this will never happen.

And so to complete the explanation. Before, we had the product_id that could be used as the deterministic element in whether the line.purchase_line_id would be in the po.order_line while now, the best deterministic element is the product.name because the line.name could not be used to match the lines so the matched product ended up changing.

If this is not the change we want to keep, it will mean modifying the matching process of the invoices and purchase_orders which i believe shouldn't be done simultaneously with this.